File: day10/day10.rs
use std::collections::HashMap;
use std::fs;

#[derive(Debug, Clone)]
struct Machine {
    target_lights: Vec<u8>,
    buttons: Vec<Vec<usize>>,
    joltages: Vec<i64>,
    num_lights: usize,
}

fn parse_input(input: &str) -> Vec<Machine> {
    input
        .lines()
        .filter(|line| !line.trim().is_empty())
        .map(|line| {
            let parts: Vec<&str> = line.split_whitespace().collect();

            // Parse indicator lights pattern [.##.]
            let lights_str = parts[0].trim_matches(|c| c == '[' || c == ']');
            let num_lights = lights_str.len();
            let target_lights: Vec<u8> = lights_str
                .chars()
                .map(|c| if c == '#' { 1 } else { 0 })
                .collect();

            // Parse buttons (0,1,2)
            let buttons: Vec<Vec<usize>> = parts[1..parts.len() - 1]
                .iter()
                .map(|button_str| {
                    button_str
                        .trim_matches(|c| c == '(' || c == ')')
                        .split(',')
                        .filter_map(|s| s.parse().ok())
                        .collect()
                })
                .collect();

            // Parse joltages {3,5,4,7}
            let joltages: Vec<i64> = parts
                .last()
                .unwrap()
                .trim_matches(|c| c == '{' || c == '}')
                .split(',')
                .filter_map(|s| s.parse().ok())
                .collect();

            Machine {
                target_lights,
                buttons,
                joltages,
                num_lights,
            }
        })
        .collect()
}

fn solve_part1(target: &[u8], buttons: &[Vec<usize>], num_lights: usize) -> Option<usize> {
    let num_buttons = buttons.len();

    // Build matrix: rows are lights, columns are buttons
    // matrix[light][button] = 1 if button affects light
    let mut matrix = vec![vec![0u8; num_buttons + 1]; num_lights];

    for (button_idx, button) in buttons.iter().enumerate() {
        for &light in button {
            if light < num_lights {
                matrix[light][button_idx] = 1;
            }
        }
    }

    // Set target values in augmented column
    for (light, &val) in target.iter().enumerate() {
        matrix[light][num_buttons] = val;
    }

    // Gaussian elimination to RREF over GF(2)
    let mut pivot_row = 0;
    let mut pivot_col_to_row = HashMap::new();

    for col in 0..num_buttons {
        if pivot_row >= num_lights {
            break;
        }

        // Find pivot in current column
        let mut row = pivot_row;
        while row < num_lights && matrix[row][col] == 0 {
            row += 1;
        }

        if row < num_lights {
            // Swap rows to bring pivot to position
            matrix.swap(pivot_row, row);

            // Eliminate all other rows using XOR (GF(2) addition)
            for r in 0..num_lights {
                if r != pivot_row && matrix[r][col] == 1 {
                    for k in col..=num_buttons {
                        matrix[r][k] ^= matrix[pivot_row][k];
                    }
                }
            }

            pivot_col_to_row.insert(col, pivot_row);
            pivot_row += 1;
        }
    }

    // Check for inconsistency
    for row in &matrix[pivot_row..num_lights] {
        if row[num_buttons] == 1 {
            return None;
        }
    }

    // Identify free variables
    let free_vars: Vec<usize> = (0..num_buttons)
        .filter(|col| !pivot_col_to_row.contains_key(col))
        .collect();

    // Try all combinations to find minimum Hamming weight
    let num_free = free_vars.len();
    let mut min_presses = usize::MAX;

    for mask in 0..(1u64 << num_free) {
        let mut solution = vec![0u8; num_buttons];

        // Set free variables based on bitmask
        for (bit_idx, &var_idx) in free_vars.iter().enumerate() {
            if (mask >> bit_idx) & 1 == 1 {
                solution[var_idx] = 1;
            }
        }

        // Back-substitute to find pivot variable values
        for (&pivot_col, &pivot_row) in &pivot_col_to_row {
            let mut val = matrix[pivot_row][num_buttons];
            for &free_var in &free_vars {
                if matrix[pivot_row][free_var] == 1 {
                    val ^= solution[free_var];
                }
            }
            solution[pivot_col] = val;
        }

        // Count button presses
        let presses = solution.iter().filter(|&&v| v == 1).count();
        min_presses = min_presses.min(presses);
    }

    Some(min_presses)
}

fn solve_part2(target: &[i64], buttons: &[Vec<usize>]) -> Option<i64> {
    let num_requirements = target.len();
    let num_buttons = buttons.len();

    // Build matrix: rows are requirements, columns are buttons
    // matrix[requirement][button] = 1 if button affects requirement
    let mut matrix = vec![vec![0i128; num_buttons + 1]; num_requirements];

    for (button_idx, button) in buttons.iter().enumerate() {
        for &req in button {
            if req < num_requirements {
                matrix[req][button_idx] = 1;
            }
        }
    }

    for (req, &val) in target.iter().enumerate() {
        matrix[req][num_buttons] = val as i128;
    }

    // Fraction-free Gaussian elimination (forward pass only)
    let mut pivot_row = 0;
    let mut pivot_col_to_row = HashMap::new();

    for col in 0..num_buttons {
        if pivot_row >= num_requirements {
            break;
        }

        // Find non-zero pivot
        let mut row = pivot_row;
        while row < num_requirements && matrix[row][col] == 0 {
            row += 1;
        }

        if row < num_requirements {
            matrix.swap(pivot_row, row);

            let pivot_val = matrix[pivot_row][col];

            // Eliminate rows below
            for r in pivot_row + 1..num_requirements {
                if matrix[r][col] != 0 {
                    let factor = matrix[r][col];
                    for k in col..=num_buttons {
                        matrix[r][k] = matrix[r][k] * pivot_val - matrix[pivot_row][k] * factor;
                    }
                }
            }

            pivot_col_to_row.insert(col, pivot_row);
            pivot_row += 1;
        }
    }

    // Check for inconsistency
    for row in &matrix[pivot_row..num_requirements] {
        if row[num_buttons] != 0 {
            return None;
        }
    }

    // Identify free variables
    let free_vars: Vec<usize> = (0..num_buttons)
        .filter(|col| !pivot_col_to_row.contains_key(col))
        .collect();

    // Search for minimum solution
    let mut min_total: Option<i64> = None;

    fn search(
        free_idx: usize,
        free_vars: &[usize],
        free_vals: &mut Vec<i64>,
        matrix: &[Vec<i128>],
        row_to_pivot_col: &HashMap<usize, usize>,
        num_buttons: usize,
        num_pivots: usize,
        min_total: &mut Option<i64>,
    ) {
        if free_idx == free_vars.len() {
            // Try to solve with current free variable values
            let mut solution = vec![0i128; num_buttons];

            // Set free variables
            for (i, &var) in free_vars.iter().enumerate() {
                solution[var] = free_vals[i] as i128;
            }

            // Back-substitute from bottom to top
            for row in (0..num_pivots).rev() {
                if let Some(&pivot_col) = row_to_pivot_col.get(&row) {
                    let pivot_val = matrix[row][pivot_col];
                    let mut rhs = matrix[row][num_buttons];

                    // Subtract known variable contributions
                    for col in pivot_col + 1..num_buttons {
                        rhs -= matrix[row][col] * solution[col];
                    }

                    // Check if solution is integral and non-negative
                    if rhs % pivot_val != 0 {
                        return; // Not integral
                    }

                    let val = rhs / pivot_val;
                    if val < 0 {
                        return; // Negative solution
                    }

                    solution[pivot_col] = val;
                }
            }

            // Calculate total button presses
            let total: i128 = solution.iter().sum();
            if total >= 0 && total < i64::MAX as i128 {
                let total_i64 = total as i64;
                if min_total.is_none() || total_i64 < min_total.unwrap() {
                    *min_total = Some(total_i64);
                }
            }
            return;
        }

        // Search range: larger for single free var, smaller for multiple
        let limit = if free_vars.len() > 1 { 200 } else { 20000 };
        for val in 0..=limit {
            free_vals[free_idx] = val;
            search(
                free_idx + 1,
                free_vars,
                free_vals,
                matrix,
                row_to_pivot_col,
                num_buttons,
                num_pivots,
                min_total,
            );
        }
    }

    // Build row -> pivot_col map once for back-substitution
    let mut row_to_pivot_col = HashMap::new();
    for (&col, &row) in &pivot_col_to_row {
        row_to_pivot_col.insert(row, col);
    }
    let num_pivots = pivot_col_to_row.len();

    let mut free_vals = vec![0i64; free_vars.len()];
    search(
        0,
        &free_vars,
        &mut free_vals,
        &matrix,
        &row_to_pivot_col,
        num_buttons,
        num_pivots,
        &mut min_total,
    );

    min_total
}

fn main() {
    let input = fs::read_to_string("input.txt").expect("Failed to read input.txt");
    let machines = parse_input(&input);

    // Part 1: Binary configuration using Gaussian elimination over GF(2)
    let part1_total: usize = machines
        .iter()
        .filter_map(|machine| {
            solve_part1(&machine.target_lights, &machine.buttons, machine.num_lights)
        })
        .sum();

    println!("Part 1: {}", part1_total);

    // Part 2: Integer linear programming with branch-and-bound search
    let part2_total: i64 = machines
        .iter()
        .filter_map(|machine| {
            if !machine.joltages.is_empty() {
                solve_part2(&machine.joltages, &machine.buttons)
            } else {
                Some(0)
            }
        })
        .sum();

    println!("Part 2: {}", part2_total);
}

File: day10/Day10.cs
using System.Text.RegularExpressions;
using System.Runtime.CompilerServices;

var machines = File.ReadAllLines("input.txt")
    .Where(line => !string.IsNullOrWhiteSpace(line))
    .Select(ParseMachine)
    .ToArray();

var part1 = 0;
foreach (var m in machines)
{
    var result = SolveGF2(m.TargetLights, m.Buttons, m.NumLights);
    if (result.HasValue)
        part1 += result.Value;
}

var part2 = 0L;
foreach (var m in machines)
{
    if (m.Joltages.Length > 0)
    {
        var result = SolveILP(m.Joltages, m.Buttons);
        if (result.HasValue)
            part2 += result.Value;
    }
}

Console.WriteLine($"Part 1: {part1}");
Console.WriteLine($"Part 2: {part2}");

(byte[] TargetLights, int NumLights, int[][] Buttons, long[] Joltages) ParseMachine(string line)
{
    var patternMatch = Regex.Match(line, @"\[([.#]+)\]");
    var pattern = patternMatch.Groups[1].Value;
    var targetLights = new byte[pattern.Length];
    for (var i = 0; i < pattern.Length; i++)
        targetLights[i] = pattern[i] == '#' ? (byte)1 : (byte)0;

    var buttonMatches = Regex.Matches(line, @"\(([0-9,]*)\)");
    var buttons = new int[buttonMatches.Count][];
    for (var i = 0; i < buttonMatches.Count; i++)
    {
        var content = buttonMatches[i].Groups[1].Value;
        if (string.IsNullOrWhiteSpace(content))
            buttons[i] = [];
        else
        {
            var parts = content.Split(',');
            buttons[i] = new int[parts.Length];
            for (var j = 0; j < parts.Length; j++)
                buttons[i][j] = int.Parse(parts[j]);
        }
    }

    var joltageMatch = Regex.Match(line, @"\{([0-9,]+)\}");
    long[] joltages;
    if (joltageMatch.Success)
    {
        var parts = joltageMatch.Groups[1].Value.Split(',');
        joltages = new long[parts.Length];
        for (var i = 0; i < parts.Length; i++)
            joltages[i] = long.Parse(parts[i]);
    }
    else
    {
        joltages = [];
    }

    return (targetLights, targetLights.Length, buttons, joltages);
}

int? SolveGF2(byte[] target, int[][] buttons, int numLights)
{
    var numButtons = buttons.Length;
    var matrix = new byte[numLights][];
    for (var i = 0; i < numLights; i++)
        matrix[i] = new byte[numButtons + 1];

    for (var buttonIdx = 0; buttonIdx < numButtons; buttonIdx++)
    {
        var btn = buttons[buttonIdx];
        for (var li = 0; li < btn.Length; li++)
        {
            var light = btn[li];
            if (light < numLights)
                matrix[light][buttonIdx] = 1;
        }
    }

    for (var light = 0; light < numLights; light++)
        matrix[light][numButtons] = target[light];

    var pivotCols = new int[numLights];
    var pivotRows = new int[numButtons];
    for (var i = 0; i < numButtons; i++) pivotRows[i] = -1;
    var pivotCount = 0;
    var currentRow = 0;

    for (var col = 0; col < numButtons; col++)
    {
        var pivotRow = -1;
        for (var r = currentRow; r < numLights; r++)
        {
            if (matrix[r][col] == 1) { pivotRow = r; break; }
        }

        if (pivotRow >= 0 && currentRow < numLights)
        {
            if (pivotRow != currentRow)
            {
                var temp = matrix[currentRow];
                matrix[currentRow] = matrix[pivotRow];
                matrix[pivotRow] = temp;
            }

            var pivotRowData = matrix[currentRow];
            for (var r = 0; r < numLights; r++)
            {
                if (r != currentRow && matrix[r][col] == 1)
                {
                    var rowData = matrix[r];
                    for (var k = col; k <= numButtons; k++)
                        rowData[k] ^= pivotRowData[k];
                }
            }

            pivotCols[pivotCount] = col;
            pivotRows[col] = currentRow;
            pivotCount++;
            currentRow++;
        }
    }

    for (var r = currentRow; r < numLights; r++)
    {
        if (matrix[r][numButtons] == 1)
            return null;
    }

    var freeVars = new int[numButtons - pivotCount];
    var freeCount = 0;
    for (var c = 0; c < numButtons; c++)
    {
        if (pivotRows[c] == -1)
            freeVars[freeCount++] = c;
    }

    var minPresses = int.MaxValue;
    var numCombinations = 1 << freeCount;
    var solution = new byte[numButtons];

    for (var mask = 0; mask < numCombinations; mask++)
    {
        Array.Clear(solution);

        for (var bitIdx = 0; bitIdx < freeCount; bitIdx++)
        {
            if (((mask >> bitIdx) & 1) == 1)
                solution[freeVars[bitIdx]] = 1;
        }

        for (var p = 0; p < pivotCount; p++)
        {
            var pivotCol = pivotCols[p];
            var row = pivotRows[pivotCol];
            var rowData = matrix[row];
            byte value = rowData[numButtons];
            for (var fi = 0; fi < freeCount; fi++)
            {
                var freeVar = freeVars[fi];
                if (rowData[freeVar] == 1)
                    value ^= solution[freeVar];
            }
            solution[pivotCol] = value;
        }

        var presses = 0;
        for (var i = 0; i < numButtons; i++)
            presses += solution[i];

        if (presses < minPresses)
            minPresses = presses;
    }

    return minPresses;
}

long? SolveILP(long[] target, int[][] buttons)
{
    var numRequirements = target.Length;
    var numButtons = buttons.Length;

    var matrix = new long[numRequirements][];
    for (var i = 0; i < numRequirements; i++)
        matrix[i] = new long[numButtons + 1];

    for (var buttonIdx = 0; buttonIdx < numButtons; buttonIdx++)
    {
        var btn = buttons[buttonIdx];
        for (var ri = 0; ri < btn.Length; ri++)
        {
            var req = btn[ri];
            if (req < numRequirements)
                matrix[req][buttonIdx] = 1;
        }
    }

    for (var req = 0; req < numRequirements; req++)
        matrix[req][numButtons] = target[req];

    var pivotCols = new int[numRequirements];
    var rowToPivotCol = new int[numRequirements];
    for (var i = 0; i < numRequirements; i++) rowToPivotCol[i] = -1;
    var colToPivotRow = new int[numButtons];
    for (var i = 0; i < numButtons; i++) colToPivotRow[i] = -1;

    var numPivots = 0;
    var currentRow = 0;

    for (var col = 0; col < numButtons; col++)
    {
        var pivotRow = -1;
        for (var r = currentRow; r < numRequirements; r++)
        {
            if (matrix[r][col] != 0) { pivotRow = r; break; }
        }

        if (pivotRow >= 0 && currentRow < numRequirements)
        {
            if (pivotRow != currentRow)
            {
                var temp = matrix[currentRow];
                matrix[currentRow] = matrix[pivotRow];
                matrix[pivotRow] = temp;
            }

            var pivotVal = matrix[currentRow][col];
            var pivotRowData = matrix[currentRow];

            for (var r = currentRow + 1; r < numRequirements; r++)
            {
                var rowData = matrix[r];
                if (rowData[col] != 0)
                {
                    var factor = rowData[col];
                    for (var k = col; k <= numButtons; k++)
                        rowData[k] = rowData[k] * pivotVal - pivotRowData[k] * factor;
                }
            }

            pivotCols[numPivots] = col;
            colToPivotRow[col] = currentRow;
            rowToPivotCol[currentRow] = col;
            numPivots++;
            currentRow++;
        }
    }

    for (var r = numPivots; r < numRequirements; r++)
    {
        if (matrix[r][numButtons] != 0)
            return null;
    }

    var freeVars = new int[numButtons - numPivots];
    var freeCount = 0;
    for (var c = 0; c < numButtons; c++)
    {
        if (colToPivotRow[c] == -1)
            freeVars[freeCount++] = c;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    long? TryBackSubstitute(long[] solution, long currentSum, long bestSoFar)
    {
        for (var p = numPivots - 1; p >= 0; p--)
        {
            var pivotCol = pivotCols[p];
            var row = colToPivotRow[pivotCol];
            var rowData = matrix[row];

            var pivotVal = rowData[pivotCol];
            var rhs = rowData[numButtons];

            for (var col = pivotCol + 1; col < numButtons; col++)
                rhs -= rowData[col] * solution[col];

            if (rhs % pivotVal != 0)
                return null;

            var value = rhs / pivotVal;
            if (value < 0)
                return null;

            solution[pivotCol] = value;
            currentSum += value;

            // Early termination: if we already exceed the best, stop
            if (bestSoFar >= 0 && currentSum >= bestSoFar)
                return null;
        }

        return currentSum;
    }

    if (freeCount == 0)
    {
        var solution = new long[numButtons];
        return TryBackSubstitute(solution, 0, -1);
    }

    var searchLimit = freeCount > 1 ? 200 : 20000;
    long minTotal = long.MaxValue;
    var solution2 = new long[numButtons];

    if (freeCount == 1)
    {
        var varIdx = freeVars[0];
        for (var value = 0L; value <= searchLimit; value++)
        {
            // Early termination: free var value already exceeds best
            if (value >= minTotal) break;

            Array.Clear(solution2);
            solution2[varIdx] = value;
            var result = TryBackSubstitute(solution2, value, minTotal);
            if (result.HasValue && result.Value < minTotal)
                minTotal = result.Value;
        }
    }
    else if (freeCount == 2)
    {
        var varIdx0 = freeVars[0];
        var varIdx1 = freeVars[1];
        for (var v0 = 0L; v0 <= searchLimit; v0++)
        {
            // Early termination
            if (v0 >= minTotal) break;

            for (var v1 = 0L; v1 <= searchLimit; v1++)
            {
                var freeSum = v0 + v1;
                if (freeSum >= minTotal) break;

                Array.Clear(solution2);
                solution2[varIdx0] = v0;
                solution2[varIdx1] = v1;
                var result = TryBackSubstitute(solution2, freeSum, minTotal);
                if (result.HasValue && result.Value < minTotal)
                    minTotal = result.Value;
            }
        }
    }
    else
    {
        // General case with pruning
        var indices = new long[freeCount];

        void Search(int depth, long currentSum)
        {
            if (currentSum >= minTotal) return;

            if (depth == freeCount)
            {
                Array.Clear(solution2);
                for (var fi = 0; fi < freeCount; fi++)
                    solution2[freeVars[fi]] = indices[fi];

                var result = TryBackSubstitute(solution2, currentSum, minTotal);
                if (result.HasValue && result.Value < minTotal)
                    minTotal = result.Value;
                return;
            }

            for (var value = 0L; value <= searchLimit; value++)
            {
                if (currentSum + value >= minTotal) break;
                indices[depth] = value;
                Search(depth + 1, currentSum + value);
            }
        }

        Search(0, 0);
    }

    return minTotal == long.MaxValue ? null : minTotal;
}
